frontend/dockerfile/linter/linter.go

```diff
@@ -5,16 +5,81 @@ import (
 	"strings"
 
 	"github.com/moby/buildkit/frontend/dockerfile/parser"
+	"github.com/pkg/errors"
 )
 
+type Config struct {
+	Warn          LintWarnFunc
+	SkipRules     []string
+	SkipAll       bool
+	ReturnAsError bool
+}
+
+type Linter struct {
+	SkippedRules  map[string]struct{}
+	CalledRules   map[string]struct{}
+	SkipAll       bool
+	ReturnAsError bool
+	Warn          LintWarnFunc
+}
+
+func New(config *Config) *Linter {
+	toret := &Linter{
+		SkippedRules: map[string]struct{}{},
+		CalledRules:  map[string]struct{}{},
+		Warn:         config.Warn,
+	}
+	toret.SkipAll = config.SkipAll
+	toret.ReturnAsError = config.ReturnAsError
+	for _, rule := range config.SkipRules {
+		toret.SkippedRules[rule] = struct{}{}
+	}
+	return toret
+}
+
+func (lc *Linter) Run(rule LinterRuleI, location []parser.Range, txt ...string) {
+	if lc.Warn == nil || lc.SkipAll {
+		return
+	}
+	rulename := rule.RuleName()
+	if _, ok := lc.SkippedRules[rulename]; ok {
+		return
+	}
+	lc.CalledRules[rulename] = struct{}{}
```

Comment:
from ci 

```
    sandbox.go:136: fatal error: concurrent map writes
    sandbox.go:136: 
    sandbox.go:136: goroutine 58 [running]:
    sandbox.go:136: github.com/moby/buildkit/frontend/dockerfile/linter.(*Linter).Run(0xc00043c5c0, {0xe219a0, 0x1389fa0}, {0xc00011b400, 0x1, 0x1}, {0xc0000345b0, 0x1, 0x1})
    sandbox.go:136: 	/src/frontend/dockerfile/linter/linter.go:48 +0xd6
    sandbox.go:136: github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb.validateBaseImagePlatform({0xc000266576, 0x6}, {{0xc00010c950, 0x5}, {0xc00010c955, 0x5}, {0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...)
    sandbox.go:136: 	/src/frontend/dockerfile/dockerfile2llb/convert.go:2313 +0x49f
```

Reply:
Kk, just updated to instead append to a slice that we run through a `map[string]struct{}` for uniqueness when we go to return the list in `Error`.